### DISCUSSION TITLE
Hetzner floating IP / static egress IP

### DISCUSSION THREAD

**Discussion by jampy** [vitobotta/hetzner-k3s#494] [General]
Following up on the discussion in issue #67 I'd like to bring up the topic again (but as an discussion, not really an issue). 

Having a static IP for egress connections is important in many scenarios.

For example, I'd like to access a Amazon RDS database in my hetzner-k3s cluster. Currently, I'd need to implement site-to-site VPN to AWS (not trivial per-se) or open up the database port to the whole Internet (obviously not a good idea) since I can't tell what public IP address my nodes will use to connect to RDS.

A simple solution, from the security perspective, would be to have a single static IP adress and configure RDS to only allow connections from that IP address.

Suggested solutions so far have been to use a dedicated (Hetzner) VM outside the cluster that acts as a gateway and uses a Hetzner floating IP address. While this would work, it defeats the purpose of a HA cluster. Once that VM goes down or has problems, all connections are gone, possibly making the whole cluster unusable.

So, I'm [thinking of a solution](https://github.com/vitobotta/hetzner-k3s/issues/67#issuecomment-1729217525) where a simple pod inside the cluster is responsible to make a Hetzner floating IP play nice with the hetzner-k3s cluster:

- upon startup, configure the floating IP address to the node the pod currently runs on (doesn't seem difficult to me)
- act itself as a forwarding gateway by sending traffic via the floating IP public address

That way, in my optimistic idea, any pod within the cluster just needs to connect to that pod (service) to ultimately reach the RDS database.

I'm lacking low-level knowledge of how Kubernetes works, so unfortunately my solution ends here, but maybe someone can help me out?

My main questions are:

- can hetzner-k3s deal with an additional public IP address?
- would it be possible to configure a known floating IP address on *all* worker nodes (so that the floating IP can be directed to any node)? Like [`sudo ip addr add 203.0.113.1 dev eth0`](https://docs.hetzner.com/cloud/floating-ips/getting-started/adding-a-floating-ip)
- what is necessary so that a pod can force egress traffic to use a specific public IP address (the floating IP), without affecting other traffic on that node?

I fear this involves messing with iptables. I hope someone can come up with something more robust.

There is a [hcloud-ip-floater](https://github.com/costela/hcloud-ip-floater) project that seems promising, but it says it needs MetalLB and as discussed with @vitobotta that is a problem as it can not easily coexist with the Hetzner load balancer. Also, I'm [not sure](https://github.com/costela/hcloud-ip-floater/issues/94) if that project helps with egress traffic.

Any ideas welcome!

**Comment by vitobotta** [comment]
If this were GCP or AWS, setting up a NAT gateway would be straightforward. However, with plain Kubernetes in Hetzner, things could get pretty complicated in my opinion.

You might consider deploying a NAT gateway pod within the cluster or using a sidecar proxy like Envoy or Squid to manage outbound traffic through a single floating IP. But it's not an easy setup, and since I haven't tried it myself, I can't provide much confidence.

In my view, it's probably better to go with an external solution, as that would be simpler to manage. But I must admit I don't have experience with this particular need, so I'm not sure what the exact steps would be.

**Comment by jampy** [comment]
Just to be sure: what do you mean by "external solution"?

**Reply by vitobotta** [reply]
You can set up regular compute instances with haproxy on each one. If you want to ensure high availability, you might use a couple of these instances along with keepalived to handle a floating IP address.

**Comment by bronhy** [comment]
How about using hezner VIP in combination with this operator https://github.com/angeloxx/cilium-haegress-operator. If i understand this correctly it  deploys a daemon that works with hetzner API to reassign the VIP to the new node. 

Or does anyone have a better solution already? 

Unfortunately my access to the DB was lost because of unexpected node restart. So i am looking for a good solution which doesn’t require a new “pet” VM
